pkg/cmd/roachtest/tests/backup.go

```diff
@@ -1103,6 +1104,369 @@ func registerBackup(r registry.Registry) {
+	err = conn.QueryRowContext(ctx, `
+		SELECT SUM((crdb_internal.range_stats(start_key)->'range_key_count')::INT)
+		FROM crdb_internal.ranges
+		WHERE database_name = 'tpch' AND table_name = 'orders'
+	`).Scan(&rangeKeys)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if rangeKeys == 0 {
+		t.Fatal("no MVCC range tombstones found")
+	}
+
+	// Fingerprint for restore comparison.
+	t.Status("fingerprinting at time 'canceled'")
+	timeCanceled := now()
+	fpCanceled, err := fingerprint(ctx, conn, "tpch", "orders")
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Status("fingerprint:\n", fpCanceled)
+	if fpInitial != fpCanceled {
+		t.Fatal("fingerprint mismatch between initial data and import cancelation")
+	}
+
+	// Now actually import the even-numbered files.
+	t.Status("importing even-numbered files")
+	if _, err = conn.ExecContext(ctx, `SET CLUSTER SETTING jobs.debug.pausepoints = ''`); err != nil {
+		t.Fatal(err)
+	}
+	err = conn.QueryRowContext(ctx, fmt.Sprintf(
+		`IMPORT INTO orders CSV DATA (%s) WITH delimiter='|', detached`, strings.Join(files, ", ")),
+	).Scan(&jobID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	waitForStatus(jobID, jobs.StatusSucceeded, "", 20*time.Minute)
+
+	// Fingerprint for restore comparison.
+	t.Status("fingerprinting at time 'completed'")
+	timeCompleted := now()
+	fpCompleted, err := fingerprint(ctx, conn, "tpch", "orders")
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Status("fingerprint:\n", fpCompleted)
+
+	// Drop the table, and wait for it to be deleted (but not GCed).
+	t.Status("dropping table")
+	if _, err = conn.ExecContext(ctx, `DROP TABLE orders`); err != nil {
+		t.Fatal(err)
+	}
+	err = conn.QueryRowContext(ctx,
+		`SELECT job_id FROM [SHOW JOBS] WHERE job_type = 'SCHEMA CHANGE GC'`).
+		Scan(&jobID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	waitForStatus(jobID, jobs.StatusRunning, sql.RunningStatusWaitingForMVCCGC, 2*time.Minute)
+
+	// Check that the data has been deleted. We don't write MVCC range tombstones
+	// unless the range contains live data, so only assert their existence if the
+	// range has any user keys.
+	var rangeID int
+	var stats string
+	err = conn.QueryRowContext(ctx, `
+		SELECT range_id, stats::STRING
+		FROM [
+			SELECT range_id, crdb_internal.range_stats(start_key) AS stats
+			FROM crdb_internal.ranges
+			WHERE database_name = 'tpch'
+		]
+		WHERE (stats->'live_count')::INT != 0 OR (
+			(stats->'key_count')::INT > 0 AND (stats->'range_key_count')::INT = 0
+		)
+	`).Scan(&rangeID, &stats)
+	if err == nil {
+		t.Fatalf("range %d not fully deleted, stats: %s", rangeID, stats)
+	} else if !errors.Is(err, gosql.ErrNoRows) {
+		t.Fatal(err)
+	}
+
+	// Take a final incremental backup.
+	t.Status("taking incremental backup")
+	err = conn.QueryRowContext(ctx,
+		`BACKUP DATABASE tpch INTO LATEST IN $1 WITH revision_history, detached`, dest).Scan(&jobID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	waitForStatus(jobID, jobs.StatusSucceeded, "", 20*time.Minute)
+
+	// Restore the initial backup.
```

Comment:
could the three restores below get put into a table driven format?

Reply:
Yeah, may as well I suppose. Will update.

Reply:
Done, PTAL.